### PR TITLE
WIP: feat: add flashcards generator with persistence, study mode, and progress tracking

### PR DESCRIPTION
**Pull request by Eziane**
> 🚧 **Draft / Work in progress** — this is **v1** of the flashcards feature. Opening as a draft to track progress and keep #204 focused on the Docker setup, as requested by @pr-elhajji. Will be marked **Ready for review** once v2 enhancements are complete. **Please do not merge yet.**

# Changelog Entry

### Description
Adds a **Flashcards generator** to the student dashboard, allowing students to extract key concepts from a tutoring session (or pasted lesson text) into a reusable revision set. Cards are generated by the configured LLM, persisted in the database, scoped per user, and come with a full study mode (3D flip animation, known/unknown marking, keyboard shortcuts, progress tracking).

**Architecture choice:** rather than re-implementing LLM connection logic in the new router, the backend calls Open WebUI's already-exposed `/api/chat/completions` endpoint, forwarding the user's Bearer token. This keeps the feature backend-agnostic — it works with any LLM Open WebUI is configured to use (Ollama, OpenAI, etc.).

**Motivation:** the platform had no post-session revision tool. A student finishing a tutoring conversation could not easily extract key points in a reusable form. This feature closes that gap.

### Added

**Backend (FastAPI):**
- `POST /api/v1/flashcards/generate` — generates flashcards from a chat session or pasted text and persists the set.
- `GET /api/v1/flashcards/sets` — lists the current user's sets.
- `GET /api/v1/flashcards/sets/{id}` — returns a single set.
- `PATCH /api/v1/flashcards/sets/{id}/progress` — updates the known-card indices.
- `DELETE /api/v1/flashcards/sets/{id}` — deletes a set.
- New SQLAlchemy model `opentutorai_flashcard_set` with fields: `id`, `user_id` (FK, scoped per user), `title`, `source_label`, `support_id` (optional link to a support), `model_used`, `cards` (JSON), `known_indices` (JSON), `created_at`, `updated_at`.
- Defensive parsing of LLM output (strips ` ```json … ``` ` markdown fences before JSON parse, since some models wrap their output).

**Frontend (SvelteKit):**
- New page at `/student/flashcards` with a two-column layout: saved sets list on the left, study area on the right.
- Generation form: model selector + choice between (a) a support tutoring session or (b) pasted lesson/conversation text.
- Study mode: 3D card flip (CSS `rotateY` + `preserve-3d`), Known / Still learning buttons, previous / next navigation.
- Per-set progress bar (known / total).
- Keyboard shortcuts: `Space` (flip), `←` / `→` (navigation), `K` (known), `U` (unknown).
- Two-step delete with 3-second visual confirmation.
- Skeleton loader while sets are loading.
- Automatic progress save (PATCH on each interaction).
- Sidebar entry **Flashcards** in the student navigation.

### Changed
- `backend/open_tutorai/main.py` — registers the new flashcards router.
- `src/lib/components/student/elements/Sidebar.svelte` — adds the Flashcards menu item.

### Deprecated
- N/A

### Removed
- N/A

### Fixed
- N/A (purely additive feature).

### Security
- All endpoints require an authenticated user; sets are strictly scoped by `user_id` so a user can only read/update/delete their own flashcards.

### Breaking Changes
**BREAKING CHANGE:** None. The new table uses the `opentutorai_` prefix to avoid colliding with Open WebUI's own tables.

---

## 🗺️ Roadmap (planned before marking Ready for review)

This is v1. The following enhancements are planned for v2 before requesting review:

- [ ] Generate flashcards from structured lesson supports (not just chat sessions).
- [ ] Export sets to PDF / CSV.
- [ ] Internationalization (i18n) of all Flashcards UI labels.
- [ ] UX polish pass (loading states, error handling, edge cases).
- [ ] Documentation update.

_Additional items may be added based on feedback. Once all roadmap items are complete and tested, this PR will be converted from Draft to Ready for review._

---

### Additional Information

**Files added:**
- `backend/open_tutorai/routers/flashcards.py`
- `backend/open_tutorai/models/database.py` (flashcard set model)
- `src/lib/apis/flashcards/index.ts` (TypeScript API client)
- `src/lib/components/student/pages/Flashcards.svelte` (main UI component)
- `src/routes/student/flashcards/+page.svelte` (SvelteKit route)

**Files modified:**
- `backend/open_tutorai/main.py`
- `src/lib/components/student/elements/Sidebar.svelte`

**Dependency on #204:** for students to see Ollama models in the selector, `BYPASS_MODEL_ACCESS_CONTROL=True` must be set in the backend env. That configuration ships in #204. Once #204 is merged, this feature works out of the box; otherwise admins need to add the variable manually or register models in the workspace.

**Notes on Svelte lifecycle:** child components' `onMount` runs before the parent layout's, so the `$models` store can briefly be empty on first render. The store is reactive and the component updates as soon as models arrive — no visible regression — but documenting it here for future readers.

### Screenshots
[Screenshots.pdf](https://github.com/user-attachments/files/27956271/Screenshots.pdf)

### CODE REVIEW

File: backend/open_tutorai/routers/flashcards.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+class FlashcardGenerateRequest(BaseModel):
+    messages: list[dict]
+    model: str
+    title: str = "Flashcard Set"
+    source_label: Optional[str] = None
+    support_id: Optional[str] = None
+
+
+class ProgressUpdateRequest(BaseModel):
+    known_indices: list[int]
```
Comment (musaabhasan) [inline backend/open_tutorai/routers/flashcards.py:77]: This needs server-side validation before persistence. `known_indices` can currently contain duplicates, negative values, or indices beyond the saved card list, which makes `known_count` and progress percentages incorrect and can leave the set in a corrupted study state. After loading the set, validate against `range(len(fs.cards or []))` and either reject invalid values with 422/400 or normalize to a de-duplicated list.

File: backend/open_tutorai/routers/flashcards.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    try:
+        content = r.json()["choices"][0]["message"]["content"].strip()
+        # Strip markdown code fences if model wrapped the JSON
+        if content.startswith("```"):
+            content = content.split("```")[1]
+            if content.startswith("json"):
+                content = content[4:]
+        data = json.loads(content.strip())
+        raw_cards = data.get("flashcards", [])
```
Comment (musaabhasan) [inline backend/open_tutorai/routers/flashcards.py]: `raw_cards` is only checked for truthiness before being committed. If the model returns a non-list, a card without `question`/`answer`, or very long/non-string values, the row can be saved and `_to_response()` will later raise when building `CardItem(**c)`, causing generate/list/get to fail. It would be safer to validate and normalize here before `db.commit()`, e.g. parse each card through the Pydantic model, enforce the 3-10 card range, and trim/limit question and answer text.

File: backend/open_tutorai/routers/flashcards.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        raise HTTPException(status_code=500, detail="Could not parse flashcards from LLM response")
+
+    db = SessionLocal()
+    try:
+        fs = FlashcardSet(
+            id=str(uuid.uuid4()),
+            user_id=user.id,
+            title=body.title,
+            source_label=body.source_label,
+            support_id=body.support_id,
```
Comment (musaabhasan) [inline backend/open_tutorai/routers/flashcards.py]: If `support_id` is meant to be source provenance, the backend should verify that the support belongs to the current user before saving it, or derive it server-side from the selected support/chat. Right now the endpoint accepts arbitrary `support_id` plus arbitrary `messages`, so a client can attach a generated set to a support it should not reference or label manual content as coming from another support.

File: src/lib/components/student/pages/Flashcards.svelte [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+	}
+
+	async function saveProgress() {
+		if (!activeSet) return;
+		savingProgress = true;
+		const token = localStorage.getItem('token') ?? '';
+		try {
+			const updated = await updateProgress(token, activeSet.id, activeSet.known_indices);
+			activeSet = { ...activeSet, updated_at: updated.updated_at };
+		} catch { /* silent — UI already updated */ }
```
Comment (musaabhasan) [inline src/lib/components/student/pages/Flashcards.svelte]: This silently keeps the optimistic UI state even when the PATCH fails, so a learner may mark cards as known, navigate away, and later find the progress was never saved. For study progress, it would be better for `saveProgress()` to return success/failure and either revert to the server state or show a visible retry/toast; `resetSet()` should also await it before presenting the reset as durable.

**Review by musaabhasan — COMMENTED** [review]
Took a pass over the flashcards draft with learner-data integrity and persistence in mind. The feature direction is useful, but I found a few issues worth addressing before marking it ready for review: validate generated card shape before saving, validate progress indices, verify source provenance, and avoid silently losing progress updates in the UI.

File: backend/open_tutorai/routers/flashcards.py (hunk 2 above)
Comment (Eziane) [reply backend/open_tutorai/routers/flashcards.py]: Jak for the feedback

**Comment by pr-elhajji** [comment]
The feature is feature-rich and mostly additive, but DB and token-forwarding concerns are important. Fixes (migration + token/config clarifications + parsing robustness + tests) will reduce risk to low/medium:

"Please add an Alembic migration for the new FlashcardSet table."
"Replace the hard-coded [http://localhost:{PORT}](http://localhost:%7BPORT%7D/) target with a configurable (env) and document it. Clarify token forwarding policy — do we want to forward user tokens or proxy with a service token?"
"Make LLM output parsing more robust (use regex to find top-level JSON) and add unit tests for parse edge cases."
"Add backend unit tests and a basic frontend test for generate + study flows. CI should run before marking Ready for review."
"Confirm table name (PREFIX + flashcard_set) is the intended name (PR description mentions opentutorai_flashcard_set)."

**Review by Dakir-Ai — CHANGES_REQUESTED** [review]
Nice work! This feature-rich and user-friendly flashcard system provides an engaging study experience.
Before merging, just make sure to add backend validation, an Alembic migration for the new table, and a few tests (as mentioned in previous review comments). Thanks for the detailed implementation and clear description!

File: backend/open_tutorai/routers/flashcards.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        r.raise_for_status()
+    except httpx.HTTPStatusError as e:
+        log.error(f"LLM call failed: {e.response.status_code} {e.response.text}")
+        raise HTTPException(status_code=502, detail="LLM call failed")
+    except Exception as e:
+        log.error(f"LLM call error: {e}")
+        raise HTTPException(status_code=502, detail="Could not reach LLM")
+
+    try:
+        content = r.json()["choices"][0]["message"]["content"].strip()
```
Comment (khadyja-R) [inline backend/open_tutorai/routers/flashcards.py]: Maybe worth adding a small guard here before accessing `choices[0].message.content`. If the upstream response shape changes a bit or returns an empty `choices` array, this could fail before the JSON parsing logic below even runs.

Comment (Eziane) [reply backend/open_tutorai/routers/flashcards.py]: Tackled

File: backend/open_tutorai/routers/flashcards.py (hunk 2 above)
Comment (Eziane) [reply backend/open_tutorai/routers/flashcards.py]: We've addressed it: there's now a _validate_flashcards() helper that runs before db.commit() and:

rejects non-list raw_cards and non-dict items;
requires question and answer to be non-empty strings (after strip());
trims question to 500 chars and answer to 1500 chars to avoid oversized rows;
parses each card through the Pydantic CardItem model so type guarantees match _to_response();
enforces the 3–10 card range and raises a 500 if the model returns outside that range.
The matching guard is also in place on the LLM response itself: choices is now checked for being a non-empty list, and choices[0].message.content is verified to be a string before parsing — so an empty/malformed completion can no longer slip through to the DB.

File: backend/open_tutorai/routers/flashcards.py (hunk 3 above)
Comment (Eziane) [reply backend/open_tutorai/routers/flashcards.py]: Tackled

**Comment by Ouissal99** [comment]
Thanks for the work on this! The feature looks solid, and the UI shown in the demo is great. I've reviewed the changes and everything seems to be heading in the right direction.
